### ISSUE TITLE
Still leaking nimsuggest process

### ISSUE THREAD

**Issue by codehz** [saem/vscode-nim#26]
![image](https://user-images.githubusercontent.com/13158903/107247990-d35c8400-6a6c-11eb-8a4a-60dc5661ec98.png)

(command: `while ($true) { get-process nimsuggest | measure | select -expand count; sleep 1 }` in pwsh

It will happens when switch to a non-nim file, and then switch back

**Comment by saem** [comment]
I'll take a quick look at the code again, but if you could dive into debugging a bit it would help a lot. Thanks.

**Comment by arkanoid87** [comment]
it happens to me too:
vscode v1.55.2
extension v0.1.20

htop shows dozens of "nimsuggest --epc --v2 relative_nim_file_path" with multiple items for the same .nim file. It becomes evident after a couple hours without restarting vscode or when opening many tabs. Sadly it won't stop until system crashes completely (I have to hard reboot). I'm currently using an alarm to remember me when ram reaches threshold.
I'll try to isolate/replicate the problem

**Comment by arkanoid87** [comment]
I'm using

```watch -n 0.2 'ps -aux | grep toolchains/nim-1.4.6/bin/nimsuggest | grep -v watch | grep -v grep'```

By opening and closing nim file tabs it shows quite immediately that nimsuggest processes spawn but are never killed and sometimes two nimsuggest processes are open on the same file

**Comment by RSDuck** [comment]
do you set a project file?

**Comment by arkanoid87** [comment]
I'm was not using "nim.project" / "nim.projectMapping".
I've set now "nim.project" to the relative path of the .nim file that isMainModule.

Documentation is unclear about what is a "project file" is, and what's a project according to extension. Should I put only the isMainModule file, or all the .nim files that are part of the project? What's the meaning of "non project mode" when using "nim.projectMapping"?

UPDATE: nevermind, I feel like I'm finding my answers reading https://nim-lang.org/docs/nimsuggest.html . Also ps now shows just a single nimsuggest process on the single nim file in nim.project

**Comment by RSDuck** [comment]
yeah, with project file mode there's only one nimsuggest instance and all requests will be processed within the context of the main file.

The default mode (one process per file) was implemented before nimsuggest supported foreign files (i.e. ones which aren't in the import/include tree of the project file).

**Comment by arkanoid87** [comment]
why not running a nimsuggest process on all .nim files in workspace? What is the difference in using "nim.project" and "nim.projectMapping"?

**Comment by RSDuck** [comment]
> why not running a nimsuggest process on all .nim files in workspace? 

you mean one nimsuggest instance which serves for all files? Because the compilation nim files can compile differently based on the context of how they're imported. So unless it's explicitly given (as the project file), this doesn't properly work. @saem also has some plans to infer this file automatically.

> What is the difference in using "nim.project" and "nim.projectMapping"?

I never used this one but I think it's like project files except on per file granularity.

**Comment by arkanoid87** [comment]
I'm now using "nim.project" settings with my isMainModule, but yet I keep getting multiple nimsuggest processes on that single file.

Surely it is no more a ram problem as before where I had dozen, but if this is not the intended behavior the process leak is real even in project mode

**Comment by saem** [comment]
I'm guessing what's happening is that nimsuggest takes a long time at some point, the extension tries to kill it or access it again, and can't, spawns a new one.

Hard to say as I can't replicate it, I have a ridiculous computer and it likely doesn't hit the issue as easily. So if someone is up for debugging it, then I think it can be diagnosed to more precisely identify the issue or issues at play.

Debugging this involves:
- clone this repo
- open it in VSCode
- press F5 (or whatever to debug)
- watch a new VSCode window open
- in the original window switch to the debug console to see console.out spam
- in the new window open your project where you have issues
- do some stuff to trigger the issue and check the console spam
- add more console spam as needed and hit the reload button in the debug overlay (it's a tiny set of buttons that popup)

Long story short, if you have VSCode and num installed you're likely two minute away from being able to see what's wrong.

**Comment by gavr123456789** [comment]
Unfortunately, the plugin is currently unused for me, a new nimsuggest processes opens when you go to each new file, each takes 500 megobytes and the system quickly begins to hang. The plugin from Konstantin seems to be dead and is not supported. Nimsuggest has terrible autocompletion and the plugin for idea has also stopped receiving commits since March (((  
So it looks like the whole language is now without the support of development tools.

**Comment by saem** [comment]
fixed

**Comment by quantimnot** [comment]
This is not fixed for me (babec4df0661d246522fb0fa76059626a62baa9f):
```
72391   ??  S      0:00.00 /Volumes/origin/ws/_wrk/var/cache/nimble/bin/nimsuggest --epc --v2 tests/effects/tdiagnostic_messages.nim
72392   ??  S      0:00.61 /Volumes/origin/ws/ports/rdn/org.nim-lang.nim/origin.src/bin/nimsuggest --epc --v2 tests/effects/tdiagnostic_messages.nim
72393   ??  S      0:00.01 /Volumes/origin/ws/_wrk/var/cache/nimble/bin/nimsuggest --epc --v2 tests/effects/tdiagnostic_messages.nim
72394   ??  S      0:00.61 /Volumes/origin/ws/ports/rdn/org.nim-lang.nim/origin.src/bin/nimsuggest --epc --v2 tests/effects/tdiagnostic_messages.nim
```
This very quickly grows until I run out of resources.
Also, it is oddly executing all `nimsuggest` executables found in `PATH`.

**Comment by quantimnot** [comment]
> Also, it is oddly executing all `nimsuggest` executables found in `PATH`.

This is just `choosenim` invoking the real `nimsuggest`.

**Comment by amedlock** [comment]
Still happening for me. Memory climbing at a constant rate.

**Comment by RSDuck** [comment]
are you sure you have a project file set?

**Comment by amedlock** [comment]
Sorry, but what does that have to do with memory leaks?  I open a folder using VSCodium and the memory starts going up.

I am not debugging, running, or even typing and this happens.

**Comment by RSDuck** [comment]
when there's no project file set, there should be one nimsuggest process per opened Nim file.

**Comment by amedlock** [comment]
That's fine but with a single open `.nim` file the one **nimsuggest** process reaches 1 GB in less than 2 minutes before I kill i.

**Comment by RSDuck** [comment]
how complex is the file you're talking about? Because nimsuggest eating huge amounts of memory (starting with somewhat larger projects) and sometimes getting into endless loops, etc. is a known issue, but there's nothing we can do about it from the extension side.

**Comment by amedlock** [comment]
50 loc. 1 import (SDL)

**Comment by RSDuck** [comment]
can you put a gist with the code here?

**Comment by amedlock** [comment]
I deleted the extension (and preferences) and re-installed. Seems good now.  Thanks.

**Comment by RSDuck** [comment]
that's odd.

**Comment by quantimnot** [comment]
My problem was a bug that I introduced in the compiler. It caused some data to be written to stdout before the port number was written out. That lead to NAN and the error handling code failed to handle it properly and just leaked it instead.

**Comment by quantimnot** [comment]
There are current reports of this problem: https://forum.nim-lang.org/t/8560

**Comment by RSDuck** [comment]
> There are current reports of this problem: https://forum.nim-lang.org/t/8560

with project file or not?

**Comment by quantimnot** [comment]
Hmm, I don't know. But I have noticed nimsuggest processes sometimes remaining in great numbers after I close all vscode sessions. I've been running multiple sessions, stop and start, for a couple weeks without a reboot. I just closed all the sessions and checked for nimsuggest procs and there were somewhere around a hundred. There were about 20 just for system.nim. I will try to debug what's happening on my system. I'm running on macos.
